### DISCUSSION TITLE
Combining vim relative line jumping with up/down display line movement

### DISCUSSION THREAD

**Discussion by timothyis** [zed-industries/zed#20863] [Feature Requests]
I'm using the following configuration to move display lines with <kbd>j</kbd> and <kbd>k</kbd> (From: https://github.com/zed-industries/zed/issues/6753#issuecomment-1938305740):
```json
{
    "context": "Editor && VimControl && !VimWaiting && !menu",
    "bindings": {
      "j": ["vim::Down", { "displayLines": true }],
      "k": ["vim::Up", { "displayLines": true }]
    }
}
```

However, this has the effect of modifying how all line jumps work. e.g., if I type <kbd>4k</kbd>, it'll move 4 display lines up, rather than going to the line that matches the 4th relative line above the current, which is how it works in vim, even if you modify how `j` and `k` work.

Is there anyway around this? I'd like to keep sole <kbd>j</kbd> and <kbd>k</kbd> presses as moving display lines, but line number + directional movement goes to the matching relative line number.

**Comment by 0x2CA** [comment]
I think you are trying to implement lazyvim keymap.

https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/keymaps.lua

```lua
-- better up/down
map({ "n", "x" }, "j", "v:count == 0 ? 'gj' : 'j'", { desc = "Down", expr = true, silent = true })
map({ "n", "x" }, "<Down>", "v:count == 0 ? 'gj' : 'j'", { desc = "Down", expr = true, silent = true })
map({ "n", "x" }, "k", "v:count == 0 ? 'gk' : 'k'", { desc = "Up", expr = true, silent = true })
map({ "n", "x" }, "<Up>", "v:count == 0 ? 'gk' : 'k'", { desc = "Up", expr = true, silent = true })
```

According to the current zed implementation, there is no way to implement this, it requires a repeat counting context, this needs to be implemented

@ConradIrwin

**Comment by ConradIrwin** [comment]
We have a `VimCount` in the context. So you could potentially do something similar:
```
[
  {
    "context": "VimControl",
    "bindings": {
      "j": ["vim::Down", { "displayLines": true }],
      "k": ["vim::Up", { "displayLines": true }]
    }
  },
  {
    "context": "VimCount",
    "bindings": {
      "j": ["vim::Down", { "displayLines": false }],
      "k": ["vim::Up", { "displayLines": false }]
    }
  }

]
```

(Note the order is important VimCount must come after VimControl in the file, because they both match on the same node).

**Reply by 0x2CA** [reply]
> We have a `VimCount` in the context. So you could potentially do something similar:
> 
> ```
> [
>   {
>     "context": "VimControl",
>     "bindings": {
>       "j": ["vim::Down", { "displayLines": true }],
>       "k": ["vim::Up", { "displayLines": true }]
>     }
>   },
>   {
>     "context": "VimCount",
>     "bindings": {
>       "j": ["vim::Down", { "displayLines": false }],
>       "k": ["vim::Up", { "displayLines": false }]
>     }
>   }
> 
> ]
> ```
> 
> (Note the order is important VimCount must come after VimControl in the file, because they both match on the same node).

Didn't you say that I really do not know

![image](https://github.com/user-attachments/assets/ade04177-c4cc-41e5-9a05-a820faa3322a)

I think this context needs to be added to the vim documentation so that those who come after are not confused

**Reply by timothyis** [reply]
Worked perfect, thank you, Conrad!

I also didn't know about the Keyboard Context view, so thank you for pointing that out @0x2CA 🙇🏼 

I did mention in the linked issue that it would be good to accept ranges in keymap bindings and I stick by that, fwiw @ConradIrwin  👀

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
